### DISCUSSION TITLE
How can obstacles be made effective in findPath calculations?

### DISCUSSION THREAD

**Discussion by liushzhql** [isaac-mason/recast-navigation-js#272] [General]
Hello,

I am currently developing a feature for ground navigation lines, where players control movement using a joystick controller. This feature involves recalculating and rendering the path from the character's current position to the target point every second. As a result, I utilize the findPath function exclusively, without the implementation of an Agent.

I have encountered an issue where obstacles added after the mesh build do not participate in the findPath calculation. Could you advise on how to resolve this problem?

**Comment by isaac-mason** [comment]
Hi @liushzhql! 

Paths should respect tilecache obstacles - I've created a new example to confirm this is working ok: https://recast-navigation-js.isaacmason.com/?path=/story/tilecache-obstacles--path-example

One thing to check is whether you've called `tileCache.update()` after adding/removing obstacles to update the navmesh: https://docs.recast-navigation-js.isaacmason.com/classes/index.TileCache.html#update

If you can share a reproduction of your issue I'm happy to take a look 🙂

**Reply by liushzhql** [reply]
I took another look and, yeah, the examples actually do work. Looks like the problem might be on my end with the asm.js file I was using. I'll dig into it a bit more myself, so no worries on your end. Thanks for the help!

**Reply by isaac-mason** [reply]
No problems at all! Let me know if I can help.

This is a bit off-topic, but out of curiosity how is the performance with the asm.js version you have compiled?

**Reply by rob-myers** [reply]
Hey @liushzhql , @isaac-mason I think I just ran into the same problem.

When using a smaller `cs` of `0.05` I had to update the tileCache many times after adding an obstacle.

```tsx
const obstacle = api.nav.tileCache.addBoxObstacle(position, extent, angle);
for (let i = 0; i < 5; i++) {
  console.log('addBoxObstacle', id, `update ${i}`);
  if (api.nav.tileCache.update(api.nav.navMesh).upToDate) break;
}
```

These updates could be spread out, to avoid blocking the main thread.

This should probably be mentioned in the README:
> https://github.com/isaac-mason/recast-navigation-js?tab=readme-ov-file#temporary-obstacles

**Reply by GGAlanSmithee** [reply]
I might also be hitting this. After adding obstacles and calling `tileCache.update`, nothing happens. I will try your method @rob-myers to see if I can brute force it to work.

What's the logic behind it working after updating the tilecache many times btw? Is it a race condition? If so, maybe the update could be done after a `timeout` instead?

**Reply by liushzhql** [reply]
@rob-myers I apologize for the delay in seeing this, but it seems the issue has already been resolved :)

**Reply by liushzhql** [reply]
@isaac-mason I'm sorry for that the performance testing hasn't been conducted yet. As I mentioned earlier, it’s just a fallback compatibility solution.
But I will share the findings immediately once available. Thanks for your understanding!

**Reply by isaac-mason** [reply]
No worries @liushzhql! I was just curious 🙂

**Comment by rob-myers** [comment]
I believe this is expected behaviour from recast/detour.

Maybe the re computation of the tile should be spread over several frames,
to avoid jank.

**Reply by GGAlanSmithee** [reply]
thanks for the info @rob-myers, this is working for me now! I used a `setInterval` without an `interval` to split it over several frames:

```ts
let times = 0;
const interval = setInterval(() => {
  tileCache.update(navMesh);
  if (times++ >= 5) clearInterval(interval);
});
```

**Reply by GGAlanSmithee** [reply]
or, more cerrectly

```ts
const interval = setInterval(() => {
  const done = tileCache.update(navMesh).upToDate;

  if (done) clearInterval(interval);
});
```

**Comment by isaac-mason** [comment]
Here's some related detour docs: https://recastnav.com/classdtTileCache.html#a93f931ad563e5cadcc389acdffd4892e

I think we can improve the documentation in our readme

> Updates the tile cache by rebuilding tiles touched by unfinished obstacle requests.
> [out] - upToDate - Whether the tile cache is fully up to date with obstacle requests and tile rebuilds. If the tile cache is up to date another (immediate) call to update will have no effect; otherwise another call will continue processing obstacle requests and tile rebuilds.

**Comment by isaac-mason** [comment]
I've made some improvements to the TileCache docs in the readme, let me know if we can make it clearer: https://github.com/isaac-mason/recast-navigation-js/commit/6b8eae7a6b89516edbb4a1058c91fa74c83f2a66

**Reply by liushzhql** [reply]
Thanks for the improvements. The documentation is much clearer now and here is a clear description of how tilecache.update works.
